src/macos/processes.rs

```diff
@@ -0,0 +1,110 @@
+
+pub fn active_executables() -> Result<ProcessList> {
+    let mut executables: HashMap<PathBuf, ProcessInfo> = HashMap::new();
+    let visible = visible_windows()?;
+    let mut sys = System::new();
+    sys.refresh_processes_specifics(ProcessRefreshKind::new());
+    for (pid, process) in sys.processes() {
+        let pid = pid.as_u32();
+        let display_name = process.name().to_string();
+        let executable = process.exe().to_path_buf();
+        let is_system = executable.starts_with("/System/");
+        match executables.entry(executable) {
+            Entry::Occupied(mut e) => {
+                let process_info = e.get();
+                if !process_info.is_visible && visible.contains(&pid) {
+                    e.get_mut().is_visible = true;
+                }
+            }
+            Entry::Vacant(e) => {
+                let executable = e.key().clone();
+                let is_visible = visible.contains(&pid);
+                e.insert(ProcessInfo {
+                    executable,
+                    display_name,
+                    is_visible,
+                    is_system,
+                });
+            }
+        }
+    }
+    Ok(executables.into_values().collect())
+}
+
+pub fn visible_windows() -> Result<HashSet<PID>> {
+    let mut pids: HashSet<PID> = HashSet::new();
+    unsafe {
+        let windows_info_list = CGWindowListCopyWindowInfo(
+            kCGWindowListOptionOnScreenOnly + kCGWindowListExcludeDesktopElements,
+            kCGNullWindowID,
+        );
+        let count = CFArrayGetCount(windows_info_list);
+
+        for i in 0..count - 1 {
+            let dic_ref = CFArrayGetValueAtIndex(windows_info_list, i);
+            let key = NSString::alloc(nil).init_str("kCGWindowOwnerPID");
+            let mut pid: *const c_void = std::ptr::null_mut();
+
+            if CFDictionaryGetValueIfPresent(
+                dic_ref as CFDictionaryRef,
+                key as *const c_void,
+                &mut pid,
+            ) != 0
+            {
+                let pid_cf_ref = pid as CFNumberRef;
+                let mut pid: i32 = 0;
+                if CFNumberGetValue(
+                    pid_cf_ref,
+                    kCFNumberSInt32Type,
+                    &mut pid as *mut i32 as *mut c_void,
+                ) {
+                    pids.insert(pid as u32);
+                }
+            }
+        }
+        Ok(pids)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn process_list() {
+        let lst = active_executables().unwrap();
+        assert!(!lst.is_empty());
+
+        for proc in &lst {
+            if !proc.is_visible {
+                dbg!(&proc.display_name);
+            }
+        }
+        dbg!(lst.len());
+    }
+
+    #[test]
+    fn visible_windows_list() {
+        let open_windows_pids = visible_windows().unwrap();
+        assert!(!open_windows_pids.is_empty());
```

Comment:
Surprised this even holds in CI, but I take it! 😃

Reply:
https://stackoverflow.com/questions/58453011/separating-real-and-dummy-windows-returned-by-cgwindowlistcopywindowinfo

this could be the reason, I have to investigate

Reply:
windows that active_windows shows are owned by specific system processes:
- window server: the process that handles windows
- control center
- TextInputMenuAgent: the process that handle the input

I tried different approaches:
- get [kCGWindowBounds](https://developer.apple.com/documentation/coregraphics/kcgwindowbounds?language=objc) field that return the Rect of the window, from the window I inspected the size and I checked if it was null 🔴
- get the [kCGWindowLayer](https://developer.apple.com/documentation/coregraphics/kcgwindowlayer?language=objc) field that returns the layer of the selected window, but it is not correlated with those processes 🔴
- get the [kCGWindowharingState](https://developer.apple.com/documentation/coregraphics/kcgwindowsharingstate?language=objc) field, but with this field we can distinguish only the windows server process 🟠

I can investigate more and find a solution in some way but I thought that most likely all the dummy processes shown, if any, would be system processes in any case, and for that reason **perhaps** we don't care much,  since whether the window `is_visible` or not should be secondary information to whether the process `is_system` or not. am I wrong? @mhils

Reply:
@emanuele-em: yes - thanks for looking into this! I think what we do here now is super good enough. 😃 👍